enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,246 @@
+
+Currently, the [cluster-dns-operator](https://github.com/openshift/cluster-dns-operator) manages DNS forwarding with a
+static ConfigMap:
+
+```bash
+$ oc get configmap/dns-default -n openshift-dns -o yaml
+apiVersion: v1
+data:
+  Corefile: |
+    .:5353 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+            pods insecure
+            upstream
+            fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        forward . /etc/resolv.conf
+        cache 30
+        reload
+    }
+kind: ConfigMap
+metadata:
+  labels:
+    dns.operator.openshift.io/owning-dns: default
+  name: dns-default
+  namespace: openshift-dns
+  <SNIP>
+```
+
+This proposal will generate this ConfigMap with a `forward` section based on user-provided values of type `Upstream`.
+
+## Motivation
+
+DNS name resolution for pods is currently provided by an instance of [CoreDNS](https://coredns.io) that runs on each
+node in the cluster. If a DNS query is for a hostname within the cluster domain suffix (i.e. `cluster.local`), CoreDNS
+resolves the name. Otherwise, CoreDNS proxies DNS messages to an upstream resolver identified by `/etc/resolv.conf` on
+the corresponding node. Although this provides a consistent and reliable approach for name resolution, it restricts how
+cluster operators manage DNS name queries.
+
+### Goals
+
+1. A well defined API for DNS forwarding configuration that is implementation agnostic.
+2. A minimal API surface that can be expanded to support future DNS forwarding use cases.
+
+### Non-Goals
+
+1. Support every possible DNS forwarding use case.
+2. Configure or manage external DNS providers.
+
+## Proposal
+
+The proposed `Upstream` type defines a schema for associating one or more `Nameservers` to a DNS subdomain identified by
+`Domain`. `Nameservers` are responsible for resolving name queries for `Domain`:
+```go
+type Upstream struct {
+	Domain string `json:"domain"`
+	Nameservers []string `json:"nameservers"`
+}
+```
+
+Each instance of CoreDNS performs health checking of `Nameservers`. If `Nameservers` consists of more than one
+nameserver, forwarding will be randomized on first use. When a healthy nameserver returns an error during the exchange,
+the next server in the list is tried. Each server is represented by an IP address or IP address and port if the server
+listens on a UDP port other than 53.
+
+A list of `Upstream` is added to `dnses.opertor.openshift.io`, allowing users the ability to configure multiple
+upstream resolvers to subdomain associations:
+
+```go
+// DNSSpec is the specification of the desired behavior of the DNS.
+type DNSSpec struct {
+	// +optional
+	Upstreams []Upstream `json:"upstreams,omitempty"`
+}
+```
+
+### User Stories
+
+#### Story 1
+
+As a customer with OpenShift running in AWS and connected to my data center by VPN, I want OpenShift DNS to resolve
+name queries for our other internal devices using the DNS servers in our data center.
```

Comment:
Does that mean that we need to be able to route requests for specific subdomains differently than other requests?  Or do we want to send all DNS traffic to the internal revolvers?

Reply:
> Does that mean that we need to be able to route requests for specific subdomains differently than other requests?

@knobunc yes, `Upstreams` will provide an interface for users to customize how CoreDNS proxies DNS messages. Whether `Upstreams` are provided or not, CoreDNS will continue to locally resolve names from the cluster domain. If no `Upstreams` are provided, CoreDNS will continue to use `/etc/resolv.conf` for names outside of the cluster domain. If an `Upstream` is provided, `Nameservers` will be used to resolve names of subdomain `Domain` and use `/etc/resolv.conf`.